src/storage/copier/node_copier.cpp

```diff
@@ -10,6 +10,52 @@ using namespace kuzu::common;
 namespace kuzu {
 namespace storage {
 
+template<typename T>
+std::unique_ptr<VirtualMorsel<arrow::Array>> CSVSharedState<T>::getMorsel() {
+    lock_t lck{this->mtx};
+    std::shared_ptr<arrow::RecordBatch> recordBatch;
+    auto result = csvStreamingReader->ReadNext(&recordBatch);
+    if (!result.ok()) {
+        throw common::CopyException(
+            "Error reading a batch of rows from CSV using Arrow CSVStreamingReader.");
+    }
+    if (recordBatch == NULL) {
+        return make_unique<CSVMorsel>(move(recordBatch), false /* failure */,
+            -1ull /* null startOffset */, -1ull /* null blockIdx */);
+    }
+    auto numRows = recordBatch->num_rows();
+    this->startOffset += numRows;
+    this->blockIdx++;
+    return make_unique<CSVMorsel>(
+        move(recordBatch), true /* success */, this->startOffset - numRows, this->blockIdx - 1);
+}
+
+template<typename T>
+std::unique_ptr<VirtualMorsel<arrow::ChunkedArray>> ParquetSharedState<T>::getMorsel() {
+    lock_t lck{this->mtx};
+    std::shared_ptr<arrow::Table> currTable;
+    if (this->blockIdx == numBlocks) {
+        return make_unique<ParquetMorsel>(move(currTable), false /* failure */,
+            -1ull /* null startOffset */, -1ull /* null blockIdx */);
+    }
+    auto result = parquetReader->RowGroup(this->blockIdx)->ReadTable(&currTable);
+    if (!result.ok()) {
+        throw common::CopyException(
+            "Error reading a batch of rows from CSV using Arrow CSVStreamingReader.");
+    }
+    // TODO(Semih): I have not verified that, similar to CSV reading, that if ReadTable runs out of
+    // blocks to read, then it sets the currTable to NULL.
+    if (currTable == NULL) {
+        return make_unique<ParquetMorsel>(move(currTable), false /* failure */,
+            -1ull /* null startOffset */, -1ull /* null blockIdx */);
```

Comment:
Ditto. For the null startOffset, you can use `INVALID_NODE_OFFSET`.

Reply:
Done.